File: src/com/grinder/game/content/skill/skillable/impl/Woodcutting.java
package com.grinder.game.content.skill.skillable.impl;

import com.grinder.game.collision.CollisionManager;
import com.grinder.game.content.achievement.AchievementManager;
import com.grinder.game.content.achievement.AchievementType;
import com.grinder.game.content.item.degrading.DegradingType;
import com.grinder.game.content.miscellaneous.PetHandler;
import com.grinder.game.content.skill.skillable.DefaultSkillable;
import com.grinder.game.content.skill.skillable.impl.woodcutting.AxeType;
import com.grinder.game.content.skill.skillable.impl.woodcutting.TreeType;
import com.grinder.game.content.skill.task.SkillMasterType;
import com.grinder.game.content.skill.task.SkillTaskManager;
import com.grinder.game.content.task_new.DailyTask;
import com.grinder.game.content.task_new.PlayerTaskManager;
import com.grinder.game.definition.ItemDefinition;
import com.grinder.game.entity.agent.player.Player;
import com.grinder.game.entity.agent.player.PlayerExtKt;
import com.grinder.game.entity.agent.player.equipment.EquipmentConstants;
import com.grinder.game.entity.grounditem.ItemOnGroundManager;
import com.grinder.game.entity.object.*;
import com.grinder.game.model.Graphic;
import com.grinder.game.model.Position;
import com.grinder.game.model.Skill;
import com.grinder.game.model.attribute.AttributeManager;
import com.grinder.game.model.item.Item;
import com.grinder.game.model.sound.Sounds;
import com.grinder.game.task.Task;
import com.grinder.game.task.TaskManager;
import com.grinder.game.task.impl.TimedObjectReplacementTask;
import com.grinder.util.ItemID;
import com.grinder.util.Misc;
import com.grinder.util.ObjectID;

import java.util.Optional;

/**
 * Represents the Woodcutting skill.
 *
 * @author Professor Oak
 */
public class Woodcutting extends DefaultSkillable {

	private static final int TREE_STUMP = 1343;
	/**
	 * The {@link GameObject} to cut down.
	 */
	private final GameObject treeObject;
	/**
	 * The {@code treeObject} as an enumerated type which contains information
	 * about it, such as required level.
	 */
	private final TreeType tree;
	/**
	 * The axe we're using to cut down the tree.
	 */
	private Optional<AxeType> axe = Optional.empty();

	/**
	 * Constructs a new {@link Woodcutting}.
	 *
	 * @param treeObject
	 *            The tree to cut down.
	 * @param tree
	 *            The tree's data
	 */
	public Woodcutting(GameObject treeObject, TreeType tree) {
		this.treeObject = treeObject;
		this.tree = tree;
	}

	@Override
	public void start(Player player) {
		player.getPacketSender().sendMessage("You swing your axe at the tree..");

		super.start(player);
	}

	@Override
	public void startAnimationLoop(Player player) {
		Task animLoop = new Task(4, player, true) {
			@Override
			protected void execute() {
				player.performAnimation(axe.get().getAnimation());

				if (Misc.random(0, 500) == 0) {
					dropBirdsNest(player);
				}
			}
		};
		TaskManager.submit(animLoop);
		getTasks().add(animLoop);
	}

	@Override
	public void startGraphicsLoop(Player player) {

	}

	@Override
	public void startSoundLoop(Player player) {
//		Task soundLoop = new Task(2, player, true) {
//			@Override
//			protected void execute() {
//				player.getPacketSender().sendSound(Sounds.HIT_TREE, 200);
//			}
//		};
//		TaskManager.submit(soundLoop);
//		getTasks().add(soundLoop);
	}

	@Override
	public void onCycle(Player player) {
		if (!ClippedMapObjects.exists(getTreeObject()) || ObjectManager.existsAt(tree.getStumpId(), treeObject.getPosition())) {
				cancel(player);
		}
	}

	private void handleTask(Player player) {
		switch(tree) {
			case NORMAL:
				PlayerTaskManager.progressTask(player, DailyTask.CHOP_LOGS);
				break;
			case WILLOW:
				PlayerTaskManager.progressTask(player, DailyTask.CHOP_WILLOW_LOGS);
				break;
			case MAPLE:
				PlayerTaskManager.progressTask(player, DailyTask.CHOP_MAPLE_LOGS);
				break;
			case YEW:
				PlayerTaskManager.progressTask(player, DailyTask.CHOP_YEW_LOGS);
				break;
			case MAGIC:
				PlayerTaskManager.progressTask(player, DailyTask.CHOP_MAGIC_LOGS);
				break;
		}
	}

	@Override
	public void finishedCycle(Player player) {
		// Add logs..
		
		
		boolean infernalAxe = (player.getEquipment().contains(13241) || player.getInventory().contains(13241)) && Misc.getRandomInclusive(4) == 1;

		if (infernalAxe && axe.get().getId() == 13241 && tree != TreeType.VINES && tree != TreeType.DRAMEN) {
			player.getSkillManager().addExperience(Skill.FIREMAKING, tree.getBurnXpReward());
			if (player.getEquipment().contains(13241)) {
			player.getItemDegradationManager().degrade(DegradingType.SKILLING, -1);
			} else {
			player.getItemDegradationManager().degradeInventoryItems(DegradingType.SKILLING, -1, 13241);	
			}
			player.performGraphic(new Graphic(86));
		} else {
			player.getInventory().add(tree.getLogId(), 1);
		}
		if (tree == TreeType.VINES)
			player.getPacketSender().sendMessage("You chop off the tangling vines.");
		else
			player.getPacketSender().sendMessage("You get some logs.");

		// Roll pet
		PetHandler.onSkill(player, Skill.WOODCUTTING);

		player.getPoints().increase(AttributeManager.Points.CHOPPED_TREES, 1); // Increase points

		if (tree == TreeType.MAGIC) {
			player.getPoints().increase(AttributeManager.Points.MAGIC_TREES_CUT, 1); // Increase points
		}
		
		// Add exp..
		handleTask(player);
		SkillTaskManager.perform(player, tree.getLogId(), 1, SkillMasterType.WOODCUTTING);
		player.getSkillManager().addExperience(Skill.WOODCUTTING, tree.getXpReward());

		/*
		 * Cutting vine
		 */
		if (tree.equals(TreeType.VINES)) {
			// Despawn original object with a temporary object
			TaskManager.submit(new TimedObjectReplacementTask(treeObject,
					DynamicGameObject.createPublic(-1, treeObject.getPosition()),
					tree.getRespawnTimer()));
			player.getMotion()
					.enqueuePathToWithoutCollisionChecks(getVineMovePosition(player).getX(), getVineMovePosition(player).getY());
			// Stop skilling..
			cancel(player);
			return;
		}

		if (tree.equals(TreeType.NORMAL)) {
			AchievementManager.processFor(AchievementType.CHOP_CHOP, player);
		} else if (tree.equals(TreeType.WILLOW)) {
			AchievementManager.processFor(AchievementType.CHOPPING_AWAY, player);
		} else if (tree.equals(TreeType.MAGIC)) {
			AchievementManager.processFor(AchievementType.TREES_ARE_LIFE, player);
			AchievementManager.processFor(AchievementType.AXE_DOES_IT, player);
		}
		// Regular trees should always despawn.
		// Multi trees are random.
		if (tree.isMulti() && Misc.getRandomInclusive(tree.getCycles() * 2) > 2) {
			return;
		}
		// Despawn original object with a temporary object (tree stump id
		// 1343)
		if (tree == TreeType.REDWOOD) {
			TaskManager.submit(new TimedObjectReplacementTask(treeObject,
					DynamicGameObject.createPublic(treeObject.getId() + 1, treeObject.getPosition(), treeObject.getObjectType(), treeObject.getFace()),
					tree.getRespawnTimer() + Misc.getRandomInclusive(3) / 2));
		} else {
			TaskManager.submit(new TimedObjectReplacementTask(treeObject,
					DynamicGameObject.createPublic(tree.getStumpId(), treeObject.getPosition(), treeObject.getObjectType(), treeObject.getFace()),
					tree.getRespawnTimer() + Misc.getRandomInclusive(3) / 2));
		}
		player.getPacketSender().sendMessage("You chop down the tree.");
		player.getPacketSender().sendAreaPlayerSound(2734);
		// Stop skilling..
		cancel(player);
		PlayerExtKt.tryRandomEventTrigger(player, 2F);
	}

	public Position getVineMovePosition(Player player) {
		if (treeObject.getPosition().equals(new Position(2690, 9564, 0)) && player.getPosition().getX() >= 2691) {
			return new Position(2689, 9564, 0);
		} else if (treeObject.getPosition().equals(new Position(2690, 9564, 0)) && player.getPosition().getX() <= 2689) {
			return new Position(2691, 9564, 0);
		}  else if (treeObject.getPosition().equals(new Position(2683, 9569, 0)) && player.getPosition().getY() <= 9568) {
			return new Position(2683, 9570, 0);
		}  else if (treeObject.getPosition().equals(new Position(2683, 9569, 0)) && player.getPosition().getY() >= 9570) {
			return new Position(2683, 9568, 0);
		}  else if (treeObject.getPosition().equals(new Position(2673, 9499, 0)) && player.getPosition().getX() >= 2674) {
			return new Position(2672, 9499, 0);
		}  else if (treeObject.getPosition().equals(new Position(2673, 9499, 0)) && player.getPosition().getX() <= 2672) {
			return new Position(2674, 9499, 0);
		}  else if (treeObject.getPosition().equals(new Position(2694, 9482, 0)) && player.getPosition().getX() >= 2695) {
			return new Position(2693, 9482, 0);
		}  else if (treeObject.getPosition().equals(new Position(2694, 9482, 0)) && player.getPosition().getX() <= 2693) {
			return new Position(2695, 9482, 0);
		}  else if (treeObject.getPosition().equals(new Position(2675, 9479, 0)) && player.getPosition().getX() >= 2676) {
			return new Position(2674, 9479, 0);
		}  else if (treeObject.getPosition().equals(new Position(2675, 9479, 0)) && player.getPosition().getX() <= 2674) {
			return new Position(2676, 9479, 0);
		}
		return treeObject.getPosition();
	}

	@Override
	public int cyclesRequired(Player player) {
		int cycles = tree.getCycles() + Misc.getRandomInclusive(5);
		cycles -= (int) player.getSkillManager().getMaxLevel(Skill.WOODCUTTING) * 0.1;
		cycles -= cycles * axe.get().getSpeed();
		if(cycles < 3) {
			cycles = 3;
		}
		return cycles;
	}
	
	@Override
	public void onCancel(Player player) {
		player.getPacketSender().sendStopSound(Sounds.HIT_TREE);
	}

	public static Optional<AxeType> getAxe(Player player) {
		Optional<AxeType> axe = Optional.empty();
		for (AxeType a : AxeType.values()) {
			if (player.getEquipment().getItems()[EquipmentConstants.WEAPON_SLOT].getId() == a.getId()
					|| player.getInventory().contains(a.getId())) {

				// If we have already found an axe,
				// don't select others that are worse or can't be used
				if (axe.isPresent()) {
					if (player.getSkillManager().getMaxLevel(Skill.WOODCUTTING) < a.getRequiredLevel()) {
						continue;
					}
					if (a.getRequiredLevel() < axe.get().getRequiredLevel()) {
						continue;
					}
				}

				axe = Optional.of(a);
			}
		}

		return axe;
	}

	@Override
	public boolean hasRequirements(Player player) {
		// Attempt to find an axe..
		axe = getAxe(player);

		// Check if we found one..
		if (!axe.isPresent()) {
			if (tree == TreeType.VINES) {
				player.getPacketSender().sendMessage("You need an axe to chop down the vines.", 1000);
			} else {
				player.getPacketSender().sendMessage("You need an axe to chop down this tree.", 1000);
				player.getPacketSender()
						.sendMessage("You do not have an axe which you have the Woodcutting level to use.", 1000);
			}
			return false;
		}

		// Check if we have the required level to cut down this {@code tree}
		// using the {@link Axe} we found..
		if (player.getSkillManager().getCurrentLevel(Skill.WOODCUTTING) < axe.get().getRequiredLevel()) {
			player.getPacketSender()
					.sendMessage("You do not have an axe which you have the Woodcutting level to use.", 1000);
			return false;
		}

		// Check if we have the required level to cut down this {@code tree}..
		if (player.getSkillManager().getCurrentLevel(Skill.WOODCUTTING) < tree.getRequiredLevel()) {
			player.getPacketSender().sendMessage(
					"You need a Woodcutting level of " + tree.getRequiredLevel() + " to chop down this tree.", 1000);
			return false;
		}

		// Finally, check if the tree object remains there.
		// Another player may have cut it down already.
		if (!ClippedMapObjects.exists(treeObject) || ObjectManager.existsAt(TREE_STUMP, treeObject.getPosition())) {
			return false;
		}

		return super.hasRequirements(player);
	}

	private void dropBirdsNest(Player player) {
		Item item = new Item(ItemID.BIRD_NEST_4);

		if (Misc.randomChance(30F)) {
			item = new Item(ItemID.BIRD_NEST_5);
		}

		Position treePos = treeObject.getPosition();
		Position randomPosition;

		for (int i = 0; i < 50; i++) {
			randomPosition = new Position(treePos.getX()-2+Misc.random(5), treePos.getY()-2+Misc.random(5), treePos.getZ());

			if (!CollisionManager.blocked(randomPosition)) {
				ItemOnGroundManager.registerNonGlobal(player, item, randomPosition);
				return;
			}

		}


	}

	@Override
	public boolean loopRequirements() {
		return true;
	}

	@Override
	public boolean allowFullInventory() {
		return false;
	}

	public GameObject getTreeObject() {
		return treeObject;
	}

}

File: src/com/grinder/game/content/skill/skillable/impl/woodcutting/WoodCutting.kt
package com.grinder.game.content.skill.skillable.impl.woodcutting

import com.grinder.game.World
import com.grinder.game.content.achievement.AchievementManager
import com.grinder.game.content.achievement.AchievementType
import com.grinder.game.content.item.degrading.DegradingType
import com.grinder.game.content.miscellaneous.PetHandler
import com.grinder.game.content.skill.skillable.DefaultSkillable
import com.grinder.game.content.skill.task.SkillMasterType
import com.grinder.game.content.skill.task.SkillTaskManager
import com.grinder.game.content.task_new.DailyTask
import com.grinder.game.content.task_new.PlayerTaskManager
import com.grinder.game.content.task_new.WeeklyTask
import com.grinder.game.definition.ItemDefinition
import com.grinder.game.entity.`object`.DynamicGameObject
import com.grinder.game.entity.`object`.GameObject
import com.grinder.game.entity.agent.player.Player
import com.grinder.game.entity.agent.player.tryRandomEventTrigger
import com.grinder.game.entity.grounditem.ItemOnGroundManager
import com.grinder.game.model.Graphic
import com.grinder.game.model.Position
import com.grinder.game.model.Skill
import com.grinder.game.model.item.Item
import com.grinder.game.task.Task
import com.grinder.game.task.TaskManager
import com.grinder.util.ItemID
import com.grinder.util.Misc

/**
 * TODO: add documentation
 *
 * @author  Stan van der Bend (https://www.rune-server.ee/members/StanDev/)
 * @since   08/04/2020
 * @version 1.0
 */
class WoodCutting(
        private val treeObject: GameObject,
        private val treeType: TreeType) : DefaultSkillable() {

    override fun startAnimationLoop(player: Player) {
        val axe = findUsableAxe(player)?:return
        val animLoop: Task = object : Task(4, player, true) {
            override fun execute() {
                player.performAnimation(axe.animation)
            }
        }
        TaskManager.submit(animLoop)
        tasks.add(animLoop)
    }

    override fun startGraphicsLoop(player: Player?) {
    }

    override fun startSoundLoop(player: Player) {}
    override fun allowFullInventory() = false
    override fun loopRequirements() = true

    override fun onCycle(player: Player) {
    }

    fun handleTask(player: Player) {
        when(treeType) {
            TreeType.NORMAL -> PlayerTaskManager.progressTask(player, DailyTask.CHOP_LOGS)
            TreeType.WILLOW -> PlayerTaskManager.progressTask(player, DailyTask.CHOP_WILLOW_LOGS)
            TreeType.MAPLE -> PlayerTaskManager.progressTask(player, DailyTask.CHOP_MAPLE_LOGS)
            TreeType.YEW -> PlayerTaskManager.progressTask(player, DailyTask.CHOP_YEW_LOGS)
            TreeType.MAGIC -> PlayerTaskManager.progressTask(player, DailyTask.CHOP_MAGIC_LOGS)
        }
    }
    /**
     * Executes when player
     * @param player - Player
     */
    override fun finishedCycle(player: Player) {

        val axe = findUsableAxe(player)

        if(axe != null) {
            if (!infernalAxeEffect(player, axe))
                player.inventory.add(treeType.logId, 1)
        }

        getAchievements().forEach {
            AchievementManager.processFor(it, player)
        }

        player.sendMessage(getChopMessage())

        SkillTaskManager.perform(player, treeType.logId, 1, SkillMasterType.WOODCUTTING)
        handleTask(player)

        player.skillManager.addExperience(Skill.WOODCUTTING, treeType.xpReward)

        if(treeType == TreeType.VINES){
            World.deSpawn(treeObject)
            TaskManager.submit(treeType.respawnTimer) {
                World.addObject(treeObject)
            }
            player.motion.traceTo(getVineMovePosition(player))
            return
        }

        if(ignoreTreeRemoval())
            return

        val trunk = DynamicGameObject(1343, treeObject, true)

        World.deSpawn(treeObject)
        World.addObject(trunk)

        TaskManager.submit(getRespawnTimer()) {
            World.deSpawn(trunk)
            World.addObject(treeObject)
        }

        player.packetSender.sendAreaPlayerSound(2734)

        cancel(player)

        player.tryRandomEventTrigger(2f)
    }


    override fun cyclesRequired(player: Player): Int {
        val axeSpeed = findUsableAxe(player)?.speed?.toInt()?:0
        var cycles = treeType.cycles + Misc.getRandomInclusive(3)
        cycles -= player.skillManager.getMaxLevel(Skill.WOODCUTTING) * 0.1.toInt()
        cycles -= cycles * axeSpeed
        if (cycles < 3)
            cycles = 1 + Misc.getRandomInclusive(2)
        return cycles
    }

    /**
     * Checks if we have Woodcutting requirements
     * @see DefaultSkillable.hasRequirements
     */
    override fun hasRequirements(player: Player): Boolean {
        var axe = findAxe(player)

        if(axe == null)
            player.sendMessage("You need an axe to chop the ${getTreeName()}", 1000)
        else {
            axe = findUsableAxe(player)
            when {
                axe == null -> player.sendMessage("You do not have an axe which you have the Woodcutting level to use.", 1000)
                !canChopTree(player) -> player.sendMessage("You need a Woodcutting level of ${treeType.requiredLevel} to chop down this tree.")
                else -> return World.isSpawned(treeObject)
                        && super.hasRequirements(player)
            }
        }
        return false
    }

    private fun infernalAxeEffect(player: Player, axeType: AxeType): Boolean {
        if(axeType == AxeType.INFERNAL && Misc.randomChance(25f)){
            player.performGraphic(Graphic(86))
            player.skillManager.addExperience(Skill.FIREMAKING, treeType.burnXpReward)

            if(player.equipment.contains(axeType.id))
                player.itemDegradationManager.degrade(DegradingType.SKILLING, -1)
            else
                player.itemDegradationManager.degradeInventoryItems(DegradingType.SKILLING, -1, axeType.id)
            return true
        }
        return false
    }

    private fun getChopMessage() = when(treeType) {
        TreeType.VINES -> "You chop off the tangling vines."
        else -> "You get some ${ItemDefinition.forId(treeType.logId).name}."
    }

    private fun findUsableAxe(player: Player) = AxeType.values().findLast {
        (player.equipment.contains(it.id) || player.inventory.contains(it.id))
                &&   player.skillManager.getMaxLevel(Skill.WOODCUTTING) >= it.requiredLevel
    }

    private fun findAxe(player: Player) = AxeType.values().findLast {
        (player.equipment.contains(it.id) || player.inventory.contains(it.id))
    }

    private fun getAchievements() = when(treeType){
        TreeType.NORMAL -> arrayOf(AchievementType.CHOP_CHOP)
        TreeType.WILLOW -> arrayOf(AchievementType.CHOPPING_AWAY)
        TreeType.MAGIC -> arrayOf(AchievementType.TREES_ARE_LIFE, AchievementType.AXE_DOES_IT)
        else -> emptyArray()
    }

    private fun getTreeName() = when(treeType) {
        TreeType.VINES -> "vines"
        else -> "trees"
    }

    private fun canChopTree(player: Player) = treeType.requiredLevel <= player.skillManager.getMaxLevel(Skill.WOODCUTTING)

    private fun getRespawnTimer() = treeType.respawnTimer + Misc.getRandomExclusive(3) / 2

    private fun ignoreTreeRemoval() = treeType.isMulti && Misc.getRandomInclusive(treeType.cycles * 2) > 2

    private fun getVineMovePosition(player: Player) = when {
        treeObject.position == Position(2690, 9564, 0) && player.position.x >= 2691 -> Position(2689, 9564, 0)
        treeObject.position == Position(2690, 9564, 0) && player.position.x <= 2689 -> Position(2691, 9564, 0)
        treeObject.position == Position(2683, 9569, 0) && player.position.y <= 9568 -> Position(2683, 9570, 0)
        treeObject.position == Position(2683, 9569, 0) && player.position.y >= 9570 -> Position(2683, 9568, 0)
        treeObject.position == Position(2673, 9499, 0) && player.position.x >= 2674 -> Position(2672, 9499, 0)
        treeObject.position == Position(2673, 9499, 0) && player.position.x <= 2672 -> Position(2674, 9499, 0)
        treeObject.position == Position(2694, 9482, 0) && player.position.x >= 2695 -> Position(2693, 9482, 0)
        treeObject.position == Position(2694, 9482, 0) && player.position.x <= 2693 -> Position(2695, 9482, 0)
        treeObject.position == Position(2675, 9479, 0) && player.position.x >= 2676 -> Position(2674, 9479, 0)
        treeObject.position == Position(2675, 9479, 0) && player.position.x <= 2674 -> Position(2676, 9479, 0)
        else -> treeObject.position
    }

}